### DISCUSSION TITLE
Does flux delete kustomization delete the resources?

### DISCUSSION THREAD

**Discussion by theFong** [fluxcd/flux2#2622] [Q&A] [answered]
Hello,
When I delete a kustomization, the resources appear to not be delete. Is this expected? In the docs it says the resources should be deleted but when I check my cluster the namespaces still exist. https://fluxcd.io/docs/cmd/flux_delete_kustomization/#examples

**Comment by kingdonb** [comment, marked as answer]
The behavior as described there depends on having `spec.prune: true` set in the Flux Kustomization.

It would be better if the docs called this out.

I opened:

* https://github.com/fluxcd/flux2/pull/2624

to address this. Hope this clarifies things!

**Comment by craigbryson** [comment]
I'm seeing something similar where - if I delete a Kustomization from my repo the kustomization object is deleted but the resources created by it remain - I have `prune: true`

I see something like this in my kustomize controller log and I was wondering if I'm perhaps seeing the error because I deleted the Kustomization object as well as the manifests that built it at the same time.

```json
{"level":"info","ts":"2022-04-07T12:09:12.299Z","logger":"controller.kustomization","msg":"Reconciliation finished in 298.74335ms, next run in 2m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"rxxxx","namespace":"flux-system","revision":"main/8bae21879b14e32055eb0f2532069dacc9415229"}
{"level":"error","ts":"2022-04-07T12:09:50.640Z","logger":"controller.kustomization","msg":"Reconciliation failed after 101.932265ms, next try in 2m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"xxxx","namespace":"flux-system","revision":"main/85324aba8f447d52ea009223d3c6ee9fec48ab22","error":"kustomization path not found: stat /tmp/xxxx/3048725746/apps/xxxx: no such file or directory"}
{"level":"info","ts":"2022-04-07T12:09:52.972Z","logger":"controller.kustomization","msg":"garbage collection completed: Kustomization/flux-system/xxxx deleted","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"fluxcd2-sync","namespace":"flux-system"}
```

It would be good to understand better how pruning is supposed to work, does the controller expect the manifest files to still exist in order to compute what should be deleted, or is some other method used like annotations on resources?

**Reply by kingdonb** [reply]
That's an interesting failure mode. Yes, I think the problem is that Kustomization never reached a ready state again, as you deleted the path that it was synchronizing from. If the path was present and empty (save maybe a `.keep` file, or some other non-YAML resource so that it would remain in Git, which does not keep empty directories) then it should be possible for the Kustomization to sync again before it is deleted, and clean up any resources rather than leave them behind. I think this was a bug in older versions of Kustomize Controller, I cannot find it now, but if you delete all resources that kustomize controller was managing, and only an empty directory remains, *it should clean up all the resources that it was managing.

If garbage collection wasn't completed and prune was enabled, I would expect that Kustomization to hang around in a broken state until somebody noticed and cleaned it up. This behavior is controlled by finalizers. (Did you notice if that Kustomization that failed to clean up after itself and had `prune: true` really was deleted, or if it was left behind stuck in `Terminating` state?)

> *it should clean up all the resources that it was managing.

The asterisk is another possible failure mode, that depends on Kubernetes own finalizers. If you delete a namespace that has resources which are needed to clean up finalizers, then you will get this other failure mode that is those resources (not only Flux resources) hanging around in Terminating state, because their Finalizers have not been cleared, and won't be cleared, after some resources were cleaned up in the wrong order.

This ordering issue is highly specific to the type of resources being managed and Flux cannot know what order to delete all resources in, so some deletes must be handled carefully.

If your intention is to remove Flux and leave all the resources in place, but clean up any Flux finalizers so the CRDs and flux namespace can be safely deleted, Flux has [flux uninstall](https://fluxcd.io/docs/installation/#uninstall) that implements the logic for cleaning up finalizers that is specific to Flux. But if your goal is to delete all resources created by Flux in the safe order, it is a bit more complicated than that, and there are a handful of failure modes we've seen repeated by users, like deleting the tenant's namespace without recognizing the service account that Kustomization uses will be the first deleted resource, in a soft multi-tenancy configuration, this is a common source of similar failure modes.

The best general advice is when you're deleting resources, make sure not to delete the Namespace when there are finalizers inside. You might at least guarantee this by ensuring namespaces are only managed outside of tenants, and never inside. That way the namespace can be deleted after a tenant has already cleaned up its own resources from inside.

**Reply by craigbryson** [reply]
Hmm according to `kubectl get kustomizations -A` the kustomization is indeed deleted. 

The flow was basically that we are decommissioning an app and a developer merged a PR which deleted the Kustomization and all of the resource manifests. That led to the log above.

I have noticed whilst investigating that a lot of our kustomizations are using `v1beta1` so I wondered if that comes into play?

**Reply by kingdonb** [reply]
This might be a bug. I think if you can reproduce this behavior, it should be reported as an issue on kustomize-controller.

A Flux Kustomization with prune enabled, that is not suspended, should not be deleted without cleaning up its resources. If you can get this issue to reproduce, please provide detailed steps to repeat it. I don't know the code well enough to say what it does exactly, you can read it and find out if you read Go, or if you have an edge case that breaks the garbage collection, we can definitely follow it up in the issue report.

**Reply by stefanprodan** [reply]
Flux should be able to delete the resources even if they are removed from Git in the same commit. Please make sure all your Kustomizations are on v1beta2, if the issue persists with the latest version of Flux, then this is a bug.

**Reply by craigbryson** [reply]
No worries- I have started to see if I can repro this with podinfo on v1beta1. I'll repeat on v1beta2 to see if anything changes and if I can reproduce successfully I'll file an issue :). It will probably be tomorrow

**Reply by craigbryson** [reply]
Raised an issue at https://github.com/fluxcd/flux2/issues/2627 - but it seems resolved by `v1beta2` although there are still some errors reported in logs
